packages/astro/src/cli/check/index.ts

```diff
@@ -1,101 +1,339 @@
-import { AstroCheck, DiagnosticSeverity } from '@astrojs/language-server';
+import { AstroCheck, DiagnosticSeverity, GetDiagnosticsResult } from '@astrojs/language-server';
 import type { AstroSettings } from '../../@types/astro';
 import type { LogOptions } from '../../core/logger/core.js';
-
 import glob from 'fast-glob';
 import * as fs from 'fs';
 import { bold, dim, red, yellow } from 'kleur/colors';
 import { createRequire } from 'module';
 import ora from 'ora';
 import { fileURLToPath, pathToFileURL } from 'url';
 import { printDiagnostic } from './print.js';
+import type { Arguments as Flags } from 'yargs-parser';
+import { debug, info } from '../../core/logger/core.js';
+import { createServer, ViteDevServer } from 'vite';
+import { createVite } from '../../core/create-vite.js';
+import type { CreateViteOptions } from '../../core/create-vite';
+import type { SyncOptions, ProcessExit } from '../../core/sync';
+import fsMod from 'fs';
 
-interface Result {
+type DiagnosticResult = {
 	errors: number;
 	// The language server cannot actually return any warnings at the moment, but we'll keep this here for future use
 	warnings: number;
 	hints: number;
-}
+};
 
-export async function check(settings: AstroSettings, { logging }: { logging: LogOptions }) {
-	console.log(bold('astro check'));
+type CheckPayload = {
+	/**
+	 * Flags passed via CLI
+	 */
+	flags: Flags;
 
-	const { syncCli } = await import('../../core/sync/index.js');
-	const syncRet = await syncCli(settings, { logging, fs });
-	// early exit on sync failure
-	if (syncRet === 1) return syncRet;
+	/**
+	 * Logging options
+	 */
+	logging: LogOptions;
+};
 
-	const root = settings.config.root;
+type CheckFlags = {
+	/**
+	 * Whether the `check` command should watch for `.astro` and report errors
+	 * @default {false}
+	 */
+	watch: boolean;
+};
 
-	const spinner = ora(` Getting diagnostics for Astro files in ${fileURLToPath(root)}…`).start();
+export enum CheckResult {
+	ExitWithSuccess,
+	ExitWithError,
+	Listen,
+}
+
+const ASTRO_GLOB_PATTERN = '**/*.astro';
+
+/**
+ * Checks `.astro` files for possible errors.
+ *
+ * If the `--check` flag is provided, the command runs indefinitely and provides diagnostics
+ * when `.astro` files are modified.
+ *
+ * Every time an astro files is modified, content collections are also generated.
+ *
+ * @param {AstroSettings} settings
+ * @param {CheckPayload} options
+ * @param {Flags} options.flags
+ * @param {LogOptions} options.logging
+ */
+export async function check(
+	settings: AstroSettings,
+	{ logging, flags }: CheckPayload
+): Promise<CheckServer> {
+	let checkFlags = parseFlags(flags);
+	let options: CreateViteOptions = { settings, logging, mode: 'build', fs };
+	if (checkFlags.watch) {
+		info(logging, 'check', 'Checking files in watch mode');
+		options.isWatcherEnabled = true;
+	} else {
+		info(logging, 'check', 'Checking files');
+	}
+	// We create a server to start doing our operations
+	const viteServer = await createServer(
```

Comment:
Just wondering, was there a specific reason a Vite server was chosen over other similar tools (chokidar or other watchers)? Does it have a better understanding of the project perhaps?

Reply:
I suggested Vite just because we already use it's file watcher for dev, so it's going to work the same, and be one fewer dep.

Reply:
Ditto! One thing that I don't like very much is that while we listen to file changes, we also run the `sync` operation, which also uses a vite server under the hoods. 

I evaluated the idea of using just one server for both operations, but that didn't turn out very well:
- content collections and check servers use different options/settings;
- I don't think it's possible to change options when a server is instantiated;

Reply:
Actually I'm with Erika here too. `createServer` takes a lot of resource to start up, it also sets up a different chokidar instance so they aren't really shared in terms of performance. Plus according to https://github.com/withastro/roadmap/discussions/502, we might also want to make `astro check` a separate package, and it would be quite heavy to rely on Vite instead of only chokidar.

So maybe it's better that we use chokidar directly for now.

Reply:
@Princesseuh @bluwy Please have another review. I just wanted to let you know that I applied your suggestion. 

> createServer takes a lot of resource to start up

Is there some article/resource/finding that I can read about it?

Reply:
> Is there some article/resource/finding that I can read about it?

It's mostly through reading what Vite does when calling `createServer()` ([code](https://github.com/vitejs/vite/blob/38ce81ceb86d42c27ec39eefa8091f47f6f25967/packages/vite/src/node/server/index.ts#L338)), which is a lot of stuff 😬